File: 프로그래머스/2/12952. N－Queen/N－Queen.java
import java.util.*;

class Solution {
    int answer = 0;
    
    public int solution(int n) {
        int[] queens = new int[n]; // queens[row] = row에 있는 column 번호
        placeQueen(n, 0, queens);
        return answer;
    }
    
    public void placeQueen(int n, int row, int[] queens){
        if(row == n) {
            answer++;
            return ;
        }
        
        for(int col=0; col<n; col++){
            if (isSafe(row, col, queens)) {
                queens[row] = col;
                placeQueen(n, row+1, queens);
            }
        }
    }
    
    public boolean isSafe(int row, int col, int[] queens){
        // 해당 위치가 가능한 지 확인
        for (int i = 0; i < row; i++) {
            if (queens[i] == col) return false; // 같은 열
            if (Math.abs(row - i) == Math.abs(col - queens[i])) return false; // 대각선
        }
        return true;
    }
}

File: 프로그래머스/2/12952. N－Queen/N－Queen.py
def solution(n):
    answer = 0
    
    ## 같은 column / 대각선에 이미 퀸이 있는지 체크용
    cols = set()       ## 사용 중인 열
    diag1 = set()      ## r-c
    diag2 = set()      ## r+c

    def dfs(row):
        nonlocal answer

        # n개의 퀸을 모두 놓은 경우
        if row == n:
            answer += 1
            return
        
        for c in range(n):
            if c in cols or (row-c) in diag1 or (row+c) in diag2:
                continue
            
            cols.add(c)
            diag1.add(row-c)
            diag2.add(row+c)

            dfs(row+1)

            cols.remove(c)
            diag1.remove(row-c)
            diag2.remove(row+c)

    dfs(0)
    return answer
